### PR TITLE
init decode code.

### PR DESCRIPTION
**Pull request by qindazhu**
Caution: the code can not be run now because of issue https://github.com/k2-fsa/snowfall/issues/31#issuecomment-738570169, we are still debugging it, just commit it now so that other guys can work on it together.

### CODE REVIEW

**Comment by qindazhu** [comment]
@danpovey you may need to run `pip install editdistance` to install dependency `editdistance`.

File: egs/librispeech/asr/simple_v1/decode.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    results = decode(dataloader=test_dl,
+                     model=model,
+                     subsampling=subsampling,
+                     device=device,
+                     LG=LG,
+                     symbols=symbol_table)
+    for ref, hyp in results:
+        print('ref=', ref, ', hyp=', hyp)
+    # compute WER
+    dist = sum(editdistance.eval(ref, hyp) for ref, hyp in results)
```
Comment (pzelasko) [inline egs/librispeech/asr/simple_v1/decode.py:163]: Use “pip install kaldialign” for identical results with Kaldi

**Comment by danpovey** [comment]
OK cool.  Did you ever get an example of a pair of word symbols that were on the same determinized arc?

**Comment by danpovey** [comment]
.. wonder if it might have been that previous problem with the EOS symbol.

File: egs/librispeech/asr/simple_v1/decode.py (hunk 1 above)
Comment (danpovey) [reply egs/librispeech/asr/simple_v1/decode.py:163]: Does it have a similar interface?  Would have to be imported as `import kaldialign`, right?

Comment (pzelasko) [reply egs/librispeech/asr/simple_v1/decode.py:163]: Yeah. See usage here https://pypi.org/project/kaldialign/

**Comment by qindazhu** [comment]
> OK cool. Did you ever get an example of a pair of word symbols that were on the same determinized arc?

Sorry, I have not debugged into it, if it's urgent, I'll do tomorrow.

**Comment by danpovey** [comment]
Thanks, Piotr!   If you're familiar with the `editdistance` package let us
know what the differences are.
Haowen: not urgent.

File: egs/librispeech/asr/simple_v1/decode.py (hunk 1 above)
Comment (pzelasko) [reply egs/librispeech/asr/simple_v1/decode.py:163]: @danpovey I am not very much familiar with `editdistance`. I only vaguely remember evaluating several libs in the past and none of them got me the same WER/alignment as Kaldi, so I ported Kaldi's code to Python myself. A pleasant side effect of `kaldialign` is that the code is way faster (I compared to `Biopython` at that time, but `editdistance` says that it's a pure Python implementation so it's likely slow too). If you want we can just compare the resulting numbers for both libs.

File: egs/librispeech/asr/simple_v1/decode.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    device_id = 1
+    device = torch.device('cuda', device_id)
+    model = Model(num_features=40, num_classes=364)
+    checkpoint = os.path.join(exp_dir, 'epoch-9.pt')
+    load_checkpoint(checkpoint, model)
+    model.to(device)
+    model.eval()
+
+    LG = LG.to(device)
+    LG.scores.requires_grad_(False)
```
Comment (pzelasko) [inline egs/librispeech/asr/simple_v1/decode.py:152]: A question out of curiosity: if we did `requires_grad_(True)` instead, would that have optimized the graph scores (i.e. the product of LM scores and pronunciation scores) alongside the acoustic model?

File: egs/librispeech/asr/simple_v1/decode.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    # wrong in determinize, we are still debugging the code
+    return
+    # load L, G, symbol_table
+    lang_dir = 'data/lang_nosp'
+    symbol_table = k2.SymbolTable.from_file(lang_dir + '/words.txt')
+
+    if not os.path.exists(lang_dir + '/LG.pt'):
+        print("Loading L_disambig.fst.txt")
+        with open(lang_dir + '/L_disambig.fst.txt') as f:
+            L = k2.Fsa.from_openfst(f.read(), acceptor=False)
```
Comment (pzelasko) [inline egs/librispeech/asr/simple_v1/decode.py:91]: Could we have comments stating what's the input and output label when the FSTs are loaded (e.g. "L input symbols: phones; output symbols: words) and when they are inverted? I think it'd be easier to follow.

File: egs/librispeech/asr/simple_v1/decode.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        print("Loading G.fsa.txt")
+        with open(lang_dir + '/G.fsa.txt') as f:
+            G = k2.Fsa.from_openfst(f.read(), acceptor=True)
+        L = k2.arc_sort(L.invert_())
+        G = k2.arc_sort(G)
+        print("Intersecting L and G")
+        LG = k2.intersect(L, G)
+        print(LG.shape)
+        print("Connect L*G")
+        LG = k2.connect(LG.invert_())
```
Comment (pzelasko) [inline egs/librispeech/asr/simple_v1/decode.py:101]: why not connect before inverting? I assume the non-(co)accessible states come from the intersection rather than inversion

File: egs/librispeech/asr/simple_v1/decode.py (hunk 1 above)
Comment (danpovey) [reply egs/librispeech/asr/simple_v1/decode.py:163]: OK.  Using kaldialign is fine with me, sounds good.

File: egs/librispeech/asr/simple_v1/decode.py (hunk 2 above)
Comment (danpovey) [reply egs/librispeech/asr/simple_v1/decode.py:152]: Mm, I doubt it would have because they are not part of the model.

File: egs/librispeech/asr/simple_v1/decode.py (hunk 3 above)
Comment (danpovey) [reply egs/librispeech/asr/simple_v1/decode.py:91]: Good idea, will try to add it at some point.  But remember, in this framework it's not really input and output symbols: it's symbols and aux_symbols.  I'd rather be consistent on this point.

File: egs/librispeech/asr/simple_v1/decode.py (hunk 4 above)
Comment (danpovey) [reply egs/librispeech/asr/simple_v1/decode.py:101]: should be equivalent, I believe.

Comment (csukuangfj) [reply egs/librispeech/asr/simple_v1/decode.py:101]: Even if it is equivalent, the fsa after `k2.connect` will contain fewer number of states and arcs, so `invert_` 
will process fewer data.

File: egs/librispeech/asr/simple_v1/decode.py (hunk 1 above)
Comment (qindazhu) [reply egs/librispeech/asr/simple_v1/decode.py:163]: `editdistance` is implemented with C++, there's a benchmark here https://github.com/roy-ht/editdistance.
 Anyway, Just mention it. I'm OK with kaldialing as well.

Comment (pzelasko) [reply egs/librispeech/asr/simple_v1/decode.py:163]: Oooh I looked up the wrong editdistance (there seem to be more than one libs with that name). Let's use whatever turns out to best.

File: egs/librispeech/asr/simple_v1/decode.py (hunk 2 above)
Comment (pzelasko) [reply egs/librispeech/asr/simple_v1/decode.py:152]: It would be sufficient to register `LG.parameters()` in the initialization of the optimizer. Assuming `Fsa` has a method called `parameters()` like `nn.Module`.

File: egs/librispeech/asr/simple_v1/decode.py (hunk 3 above)
Comment (pzelasko) [reply egs/librispeech/asr/simple_v1/decode.py:91]: Got it. In that case, it would be super useful if you had a documentation page that presents a mapping between OpenFST concepts and K2 concepts (maybe you have it already).

**Comment by pzelasko** [comment]
I applied the suggestions from the discussion here in #35, we can close this if we merge the other one.
